apps/src/aichat/redux/aichatRedux.ts

```diff
@@ -374,8 +375,15 @@ export const submitChatContents = createAsyncThunk(
   async (newUserMessageText: string, thunkAPI) => {
     const dispatch = thunkAPI.dispatch as AppDispatch;
     const state = thunkAPI.getState() as RootState;
-    const {savedAiCustomizations: aiCustomizations, chatEventsCurrent} =
-      state.aichat;
+    const {chatEventsCurrent, savedAiCustomizations} = state.aichat;
+    const levelSystemPrompt = (
+      state.lab.levelProperties as AichatLevelProperties | undefined
+    )?.aichatSettings?.levelSystemPrompt;
+    const aiCustomizations = {
+      ...savedAiCustomizations,
+      systemPrompt:
```

Comment:
Yeah, good questions!
My thought was that since it's just the part of the system prompt that is hidden from the student, to prepend it here and keep things simple.

Another option could be to add `levelSystemPrompt` to type `AichatContext` since `levelId` is already a field and this prompt is set per level. Then the level system prompt can be added on the backend.

A question I have is how the safety-checking of student system prompt/retrievals will be changed once a decision is made. Currently, done when student updates customizations but wondering if we'll set up a separate API endpoint instead. But I don't think this will impact the level system prompt.

I can send level system prompt separately to the backend if you prefer as I don't have a strong preference. But keeping separate SGTM!

Reply:
Not sure I'm exactly following, but I was imagining that JS code doesn't need the `levelSystemPrompt` at all, and we could just grab it in the controller if we have the level ID. I think the current approach is simple enough though!

Reply:
+1, I was thinking that `levelSystemPrompt` would be an additional, separate parameter provided to the aichat controller endpoints (maybe as part of the `aichatContext` argument). That way we just have a little more abstraction behind how it gets prepended with the student system prompt. I know that right now it's just as simple as prepending it to the system prompt, but if our system prompt logic gets more complicated on the backend (i.e. if we do end up including a global system prompt as well) we may want more control over how the various components get combined.

Reply:
yeah - I was thinking about adding as part of `aichatContext` as well - https://github.com/code-dot-org/code-dot-org/pull/60549#discussion_r1727541894. I can go ahead and update - thanks!

Reply:
Thought of something... in the `AichatRequest` - we have columns for `level_id`, `script_id`, and `project_id` from `aichatContext` along with `model_customizations` which includes the student system prompt.

We could:
- prepend level system prompt to student system prompt before storing in table,
- add another column called `level_properties` and store level system prompt there and prepend later in `AichatSagemakerHelper`,
- grab level system prompt in `AichatSagemakerHelper` when since we'll have `level_id`. (I think this is what you were thinking, @bencodeorg )

The last option seems like the least plumbing change at this point so I'll update that way. Thanks to you both!